[flang] FDATE extension implementation: get date and time in ctime format

yiwu0b11 opened the pull request:
reference to gfortran fdate https://gcc.gnu.org/onlinedocs/gfortran/FDATE.html
usage:
```fortran
CHARACTER(32) :: time
CALL fdate(time)
WRITE(*,*) time
```

fdate is used in the ECP proxy application https://proxyapps.exascaleproject.org/app/minismac2d/
https://github.com/Mantevo/miniSMAC/blob/f90446714226eeef650b78bce06ca4967792e74d/ref/smac2d.f#L1570

yiwu0b11 committed:
FDATE extension implementation: get date and time in ctime format

reference to gfortran fdate https://gcc.gnu.org/onlinedocs/gfortran/FDATE.html
usage:
CHARACTER(32) :: time
CALL fdate(time)
WRITE(*,*) time

DavidTruby commented on flang/runtime/command.cpp:
```diff
@@ -14,6 +14,7 @@
 #include "flang/Runtime/descriptor.h"
 #include <cstdlib>
 #include <limits>
+#include <time.h>
```
We should use the C++ headers for this (i.e. `<ctime>` in this case)
This will mean prefixing all the C functions with `std::`, e.g. `std::strchr`, `std::time_t`, `std::time` etc

DavidTruby commented on flang/runtime/command.cpp:
```diff
@@ -125,6 +126,33 @@ static bool FitsInDescriptor(
 
+void removeNewLine(char *str) {
+  char *newlinePos = strchr(str, '\n');
+  if (newlinePos != NULL) {
```
Use nullptr instead of NULL.
Alternatively in this case you can just do:
`if (newlinePos) {`. This is idiomatic in C/C++ for checking pointer validity

DavidTruby commented on flang/runtime/command.cpp:
```diff
@@ -125,6 +126,33 @@ static bool FitsInDescriptor(
+  time_t current_time;
+  time(&current_time);
+
+  char *time_string = ctime(&current_time);
```
`ctime` isn't thread safe, but I'm not sure there's a cross platform C function that is; glibc has `ctime_r` but I don't think that exists on Windows or MacOS. I also think we can't depend on the non-header parts of the C++ standard library so we likely can't use `<chrono>`, which _is_ thread safe. Someone else might have a better idea of what to do here. @jeanPerier is my recollection about usage of the C++ standard library correct here?

DavidTruby replied:
Actually, it looks like C11 added a `ctime_s` function that is thread safe that should be available on all platforms. Can we use that instead?

PAX-12-WU replied:
Done. Yes, ctime_s for windows, and ctime_r for linux/macos

DavidTruby reviewed (commented):
Thanks for the patch; everything looks good (modulo my small comments) other than the thread safety issue. I'd prefer if we could make this thread safe but I'm not sure exactly how to approach that

yiwu0b11 committed:
make fdate thread safe
ctime_s is defined in MS, ctime_r is defined in linux/macos

DavidTruby reviewed (commented):
Thanks, this is looking good, I just have a couple more small comments

DavidTruby commented on flang/runtime/command.cpp:
```diff
@@ -13,8 +13,25 @@
+#else
+inline const char *ctime_alloc(
+    char *buffer, size_t bufsize, const time_t cur_time) {
+  const char *res = ctime_r(&cur_time, buffer);
```
Never mind, POSIX_C_SOURCE doesn't exist on macos.... Isn't cross platform development fun? Keep it as-is

PAX-12-WU replied:
Hope someone in the future create a standard that works like a standard

yiwu0b11 committed:
remove new line can be hardcoded, since we know where it is

PAX-12-WU commented on flang/runtime/command.cpp:
```diff
@@ -125,6 +142,36 @@ static bool FitsInDescriptor(
+}
+
+std::int32_t RTNAME(FDate)(const Descriptor *value, const Descriptor *errmsg) {
+  FillWithSpaces(*value);
```
Yes indeed, not a error but I believe something wrong with the char

```
If input char length is >26

Without FillWithSpaces
 Fri Nov 10 15:45:25 202  �
 
With FillWithSpaces
 Fri Nov 10 15:50:53 2023     
```

PAX-12-WU replied:
never mind, it won't be a problem when I move in to `extension.cpp`. (have it implemented there, instead of calling a runtime function). and its probably the place where it should be. It is a `GNU extension` rather than a `standard extension`.
https://gcc.gnu.org/onlinedocs/gfortran/FDATE.html

yiwu0b11 committed:
move to extension.cpp

yiwu0b11 committed:
Merge branch 'main' of https://github.com/llvm/llvm-project into fdate

DavidTruby reviewed (approved):
LGTM, maybe wait for @jeanPerier to check as well though. Thanks!

yiwu0b11 committed:
emit fortran runtime error on failure, fill in spac

If the length is too short to fit completely, blank return.
If length if larger than it requires(24), fill the rest of buffer space.
hange the return type of `ctime_alloc` from char * to void, because we
don't need the return value.

PAX-12-WU commented:
update based on: 
https://github.com/llvm/llvm-project/pull/70917#discussion_r1393878064 Emit a Fortran runtime failure (instead of `assert`)
https://github.com/llvm/llvm-project/pull/70917#discussion_r1393905238 Space filling

`fdate` now produce the same result on flang, compare to fortran, where
If the length is too short to fit completely, blank return.
```fortran
  character(20) :: string
  call fdate(string)
  write(*, *) string, "X"
```
```bash
$ ../build-release/bin/flang-new test.f90 
$ ./a.out 
                      X
```
If length if larger than it requires(24), fill the rest of buffer space.
```fortran
  character(30) :: string
  call fdate(string)
  write(*, *) string, "X"
```
```bash
$ ../build-release/bin/flang-new test.f90 
$ ./a.out 
 Wed Nov 15 16:59:13 2023      X
```
The length value is hardcoded, because:
```c++
  // Day Mon dd hh:mm:ss yyyy\n\0 is 26 characters, e.g.
  // Tue May 26 21:51:03 2015\n\0
```

Commit:
Merge branch 'main' into fdate

Commit:
use copyBufferAndPad instead of str.fill and std::memcpy

Take copyBufferAndPad out of anonymous namespace and declear
in header file.

jeanPerier commented on flang/runtime/extensions.cpp:
```diff
@@ -10,9 +10,26 @@
+#else
+inline void ctime_alloc(char *buffer, size_t bufsize, const time_t cur_time,
+    Fortran::runtime::Terminator terminator) {
+  const char *res = ctime_r(&cur_time, buffer);
```
Isn't some POSIX macros required here to use ctime_r?: https://linux.die.net/man/3/ctime_r

You can define a fallback that sets an empty string '/0'.

jeanPerier replied:
LGTM, thanks

jeanPerier reviewed (approved):
Thanks, please add the POSIX macros to use ctime_r, and looks great otherwise.

yiwu0b11 committed:
add POSIX macros check and fallback

yiwu0b11 committed:
brace-initialization

yiwu0b11 committed:
add proper tests

yiwu0b11 committed:
Revert "use copyBufferAndPad instead of str.fill and std::memcpy" and use CopyAndPad from character.h instead

This reverts commit 344cbdc99ffef78702c11fb6c0c76ef35d198524.

yiwu0b11 committed:
clang format

yiwu0b11 committed:
Merge remote-tracking branch 'yi/main' into yi_dev_fdate

yiwu0b11 committed:
rebase on main

yiwu0b11 committed:
clang-format and use std::byte for char arg

yiwu0b11 committed:
use char[] instead of std::arrray<char>

yiwu0b11 committed:
correct test

yiwu0b11 committed:
add macro check for test

yi-wu-arm commented:
Hi @klausler , any thought of this patch? Thank in advance!

klausler commented on flang/include/flang/Runtime/extensions.h:
```diff
@@ -22,6 +22,9 @@ extern "C" {
 void FORTRAN_PROCEDURE_NAME(flush)(const int &unit);
 
+// GNU extension subroutine FDATE
+void FORTRAN_PROCEDURE_NAME(fdate)(std::byte *string, std::int64_t length);
```
Why `std::byte` here instead of the usual `char`?

yi-wu-arm replied:
it was `std::int8_t` but when I was doing for getlog https://github.com/llvm/llvm-project/pull/74628#discussion_r1420503257, pointed out that it would be better to use std::byte since I have to cast them into `char` anyways.

yi-wu-arm replied:
if it is `char` then it won't work for some reason

klausler replied:
Like what?

yi-wu-arm replied:
its my problem, I design the test wrong, it can be `char`, I will change it in the next commit

yi-wu-arm replied:
should I change `getlog` and `getarg` as well?

klausler commented on flang/runtime/extensions.cpp:
```diff
@@ -43,6 +66,26 @@ void FORTRAN_PROCEDURE_NAME(flush)(const int &unit) {
+  std::time(&current_time);
+  CtimeBuffer(str, sizeof(str), current_time, terminator);
+
+  // Pad space on \n\0 as well, start at index 24 (included).
```
What does "Pad space on \n\0 as well" mean?

yi-wu-arm replied:
`ctime` returns a string that follow the format:
`Day Mon dd hh:mm:ss yyyy\n\0`: `Tue May 26 21:51:03 2015\n\0`, 26 characters in total, the last two character, new line and null-terminator will be pad with space as well

klausler replied:
Do you mean that the last two bytes will be replaced with spaces?

yi-wu-arm replied:
yeah, let me rephase the comment.

Commit:
use `char` instead of `std::byte`, and better comment

yiwu0b11 committed:
test fixes

yiwu0b11 committed:
add more comment

yi-wu-arm commented:
Hi @jeanPerier I have made some changes, mainly on test cases and use `char *` instead of `std::int_8t` for arg input. Saw your activity on GitHub, could you review the changes? Thanks in advance!

jeanPerier reviewed (approved):
LGTM
